File: frontend/src/app/features/user-management/services/relationship.service.ts
import { inject, Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import {
    AddMemberRequest,
    CreateRelationshipRequest,
    Relationship,
    RelationshipMember,
    UpdateMemberRoleRequest,
    UpdateRelationshipRequest,
} from '../../../core/interfaces/relationship.dtos';
import { HttpClientService } from '../../../services/http-client.service';

@Injectable({
    providedIn: 'root',
})
export class RelationshipService {
    private readonly httpClientService = inject(HttpClientService);

    private readonly baseUrl = '/relationships';

    /** Get all relationships a specific user is a member of (admin only). */
    public getRelationshipsByUserId(userId: string): Observable<Relationship[]> {
        return this.httpClientService.get(`${this.baseUrl}/user/${userId}`);
    }

    public getRelationshipById(id: string): Observable<Relationship> {
        return this.httpClientService.get(`${this.baseUrl}/${id}`);
    }

    public createRelationship(
        payload: CreateRelationshipRequest
    ): Observable<Relationship> {
        return this.httpClientService.post(this.baseUrl, payload);
    }

    public updateRelationship(
        id: string,
        payload: UpdateRelationshipRequest
    ): Observable<Relationship> {
        return this.httpClientService.put(`${this.baseUrl}/${id}`, payload);
    }

    public deleteRelationship(id: string): Observable<void> {
        return this.httpClientService.delete(`${this.baseUrl}/${id}`);
    }

    public getMembers(id: string): Observable<RelationshipMember[]> {
        return this.httpClientService.get(`${this.baseUrl}/${id}/members`);
    }

    public addMember(
        id: string,
        payload: AddMemberRequest
    ): Observable<RelationshipMember> {
        return this.httpClientService.post(`${this.baseUrl}/${id}/members`, payload);
    }

    public updateMemberRole(
        id: string,
        memberId: string,
        payload: UpdateMemberRoleRequest
    ): Observable<RelationshipMember> {
        return this.httpClientService.put(
            `${this.baseUrl}/${id}/members/${memberId}/role`,
            payload
        );
    }

    public removeMember(id: string, memberId: string): Observable<void> {
        return this.httpClientService.delete(
            `${this.baseUrl}/${id}/members/${memberId}`
        );
    }
}

File: backend/src/main/java/com/nguyen_trong_nhat/you_and_i/features/relationship/service/RelationshipService.java
package com.nguyen_trong_nhat.you_and_i.features.relationship.service;

import com.nguyen_trong_nhat.you_and_i.features.relationship.dto.*;

import java.util.List;
import java.util.UUID;

public interface RelationshipService {

    /** Create a new relationship. The authenticated user becomes the OWNER. */
    RelationshipDTO createRelationship(CreateRelationshipRequest request);

    /** Get all relationships the authenticated user is an active member of. */
    List<RelationshipDTO> getMyRelationships();

    /** Get all relationships a specific user is a member of (admin only). */
    List<RelationshipDTO> getRelationshipsByUserId(UUID userId);

    /** Get details of a specific relationship (authenticated user must be a member). */
    RelationshipDTO getRelationshipById(UUID relationshipId);

    /** Update relationship info (OWNER or ADMIN only). */
    RelationshipDTO updateRelationship(UUID relationshipId, UpdateRelationshipRequest request);

    /** Delete a relationship entirely (OWNER only). */
    void deleteRelationship(UUID relationshipId);

    /** List all active members of a relationship (must be a member to view). */
    List<RelationshipMemberDTO> getMembers(UUID relationshipId);

    /** Add a new member to a relationship (OWNER or ADMIN only). */
    RelationshipMemberDTO addMember(UUID relationshipId, AddMemberRequest request);

    /** Change a member's role (OWNER or ADMIN only). */
    RelationshipMemberDTO updateMemberRole(UUID relationshipId, UUID memberId, UpdateMemberRoleRequest request);

    /** Remove a member from a relationship (OWNER/ADMIN, or the member themselves). */
    void removeMember(UUID relationshipId, UUID memberId);
}

